fix #22606: gammasimp (y|func(y)) + gamma(x))

smichr opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

#22606 

#### Brief description of what is fixed or changed

One~~Two~~ issue resolved:
1) divergence of behavior for `x + gamma(y)` vs `f(x) + gamma(y)`
~~2) returning of `(x + gamma(x + 1))/x` instead of `1 + gamma(x)` for `gammasimp(1 + gamma(x))`; the inverse expression is also handled~~

#### Other comments

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* simplify
    * the presence of UndefinedFunction no longer affects how expressions with gamma functions are simplified
<!-- END RELEASE NOTES -->

smichr committed:
fix #22606: gammasimp (y|f(y)) + gamma(x))**(1|-1)

smichr commented on sympy/simplify/gammasimp.py:
```diff
@@ -139,6 +150,29 @@ def gamma_factor(x):
+        if level == -1:
+            n, d = fraction(expr, exact=True)
+            def distr(d, n):
+                # divide and Add arg in n by symbolic d and return
```
```suggestion
                # divide an Add arg in n by symbolic d and return
```

smichr committed:
revert level=-1 step

smichr committed:
watch for all non-gamma/binomials

smichr committed:
not bound but rewritable as gamma

smichr committed:
fix wester Y2

smichr commented on sympy/integrals/tests/test_integrals.py:
```diff
@@ -1558,7 +1558,7 @@ def test_issue_15285():
     assert integrate(x**n * exp(-x) * log(x), (x, 0, oo)).gammasimp() == Piecewise(
         (gamma(n + 1)*polygamma(0, n) + gamma(n + 1)/n, re(n) + 1 > 0),
-        (Integral(x**n*exp(-x)*log(x), (x, 0, oo)), True))
+        (Integral(x**n*exp(-x)*log(x), (x, 0, oo)), True)),integrate(x**n * exp(-x) * log(x), (x, 0, oo)).gammasimp()
```
```suggestion
        (Integral(x**n*exp(-x)*log(x), (x, 0, oo)), True))
```

MishraAditya12126 commented:
Sir i would love to work on this issue under the guidance of yours , if no one else is assigned do assign it to me @smichr

smichr committed:
rule does not apply after rewrite(gamma)

smichr committed:
maintain expectations

smichr committed:
handle side effect expectation in caller

smichr committed:
handle no-gamma state earlier

smichr committed:
no factoring for joint_rv test

smichr commented:
> Sir i would love to work on this issue

The work is this PR. If you want to review it or have a improved proposal, that would be fine.

hanspi42 commented:
@smichr: My PR that made me submit this issue has been merged; this means you will need to change 

https://github.com/sympy/sympy/blob/a3e0f9c810cb78d8ca68b128757551783564f218/sympy/integrals/tests/test_transforms.py#L41

back to

https://github.com/sympy/sympy/blob/65fd07d897ad1249bab6ff19de3d701a01fc65bc/sympy/integrals/tests/test_transforms.py#L41

smichr committed:
sort expressions when replacing with Dummy

If signsimp is ever used during processing of an expression in which
Dummy has been used to replace expressions, the expressions must be
assigned Dummy symbols in canonical order or else expressions that were
in one order might differ from the order they would be in when replaced
with Dummy. e.g.

    >>> eq = 1/(cos(x) - cos(y))
    >>> assert signsimp(eq) == eq
    >>> d1,d2 = Dummy(), Dummy()
    >>> assert signsimp(1/(d1-d2)) == -signsimp(1/(d2-d1))

So if eq is to remain unchanged after signsimp (and d1 and d2 are going
to replace the cos terms) then the order of assigment has to match the
ordering of the cos terms.
